File: 1159-smallest-subsequence-of-distinct-characters/smallest-subsequence-of-distinct-characters.java
class Solution {
    public String smallestSubsequence(String s) {

        HashMap<Character, Integer> map = new HashMap<>();
        Set<Character> set = new HashSet<>();
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < s.length(); i++) {
            map.put(s.charAt(i), i);
        }

        for (int i = 0; i < s.length(); i++) {

            char ch = s.charAt(i);

            if (set.contains(ch)) {
                continue;
            }

            while (sb.length() > 0 &&
                   sb.charAt(sb.length() - 1) > ch &&
                   map.get(sb.charAt(sb.length() - 1)) > i) {

                char last = sb.charAt(sb.length() - 1);

                sb.deleteCharAt(sb.length() - 1);
                set.remove(last);
            }

            sb.append(ch);
            set.add(ch);
        }

        return sb.toString();
    }
}

File: 1159-smallest-subsequence-of-distinct-characters/smallest-subsequence-of-distinct-characters.cs
public class Solution {
    public string SmallestSubsequence(string s) {
        Stack<char> stack1=new Stack<char>();
        Stack<char> stack2=new Stack<char>();
        StringBuilder sb=new StringBuilder();
        for(int i=s.Length-1;i>=0;i--){
           stack1.Push(s[i]);
        }
        while(stack1.Count > 0){
          char ch = stack1.Pop();
          if(stack2.Contains(ch)){
               continue;
          }
          while(stack2.Count>0 && stack2.Peek()>ch && stack1.Contains(stack2.Peek())) 
          stack2.Pop();

          stack2.Push(ch);
        }
       char[] c=stack2.ToArray();
       for(int i=c.Length-1;i>=0;i--){
        sb.Append(c[i]);
       }
        return sb.ToString();
    }
}